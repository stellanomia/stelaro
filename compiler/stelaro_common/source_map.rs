pub struct SourceMap {
    src: String
}

impl SourceMap {
    pub fn new() -> Self {
        todo!()
    }
}