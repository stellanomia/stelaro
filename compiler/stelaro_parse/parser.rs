use crate::stelaro_ast::token::{Token, TokenStream};

use super::errors::ParseError;


pub struct Parser {
    token_stream: TokenStream,
    token: Token,
    prev_token: Token,
}

impl Parser {
    pub fn new(
        token_stream: TokenStream,
    ) -> Self {
        let mut parser = Parser {
            token_stream,
            token: Token::dummy(),
            prev_token: Token::dummy(),
        };

        parser.bump();

        parser
    }

    pub fn bump(&mut self) {

    }

    pub fn expect(&mut self, token: &Token) -> Result<(), ParseError> {
        todo!()
    }
}