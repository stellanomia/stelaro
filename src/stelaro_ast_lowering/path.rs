use crate::stelaro_ast::{ast, NodeId};
use crate::stelaro_ast_lowering::LoweringContext;
use crate::stelaro_sir::{sir, Res};


impl<'sir> LoweringContext<'_, 'sir> {
    pub fn lower_path(
        &mut self,
        id: NodeId,
        path: &ast::Path,
    ) -> sir::Path<'sir> {
        let res = self.get_res(id).unwrap_or(Res::Err);
        let res = self.lower_res(res);
        sir::Path {
            span: path.span,
            res,
            segments: self.arena.alloc_from_iter(
                path.segments.iter().map(|segment| {
                    self.lower_path_segment(segment)
                })
            ),
        }
    }

    pub fn lower_path_segment(
        &mut self,
        segment: &ast::PathSegment,
    ) -> sir::PathSegment {
        let res = self.expect_res(segment.id);
        let sir_id = self.lower_node_id(segment.id);

        sir::PathSegment {
            ident: segment.ident,
            sir_id,
            res: self.lower_res(res),
        }
    }
}
