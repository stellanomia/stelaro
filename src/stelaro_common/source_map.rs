use std::io;
use std::rc::Rc;
use std::{cell::RefCell, fs, hash::{DefaultHasher, Hash, Hasher}, path::{Path, PathBuf}};

pub struct SourceMap {
    // TODO: 単一ファイルでコードが評価出来たら複数ファイル対応(files: SourceMapFilesに変更)
    file: RefCell<Rc<SourceFile>>,
    file_loader: FileLoader
}

impl SourceMap {
    pub fn new() -> Self {
        SourceMap {
            file: Default::default(),
            file_loader: FileLoader,
        }
    }

    pub fn load_file(&self, path: &Path) -> io::Result<Rc<SourceFile>> {
        let src = self.file_loader.read_file(path)?;
        let filename = path.to_owned();
        Ok(self.new_source_file(filename, src))
    }

    pub fn new_source_file(&self, path: PathBuf, src: String) -> Rc<SourceFile> {
        // 複数ファイルの場合、filesにこのpathから得たFileIdが存在しないか確認する
        // 存在する場合、filesからRc<SourceFile>を取得し、
        // 存在しない場合、SourceFile::new()し、filesにregisterする。
        Rc::new(SourceFile::new(path, src))
    }
}

impl Default for SourceMap {
    fn default() -> Self {
        Self::new()
    }
}


struct FileLoader;

impl FileLoader {
    pub fn file_exists(&self, path: &Path) -> bool {
        path.exists()
    }

    fn read_file(&self, path: &Path) -> io::Result<String> {
        fs::read_to_string(path)
    }
}

#[derive(Debug, Default)]
pub struct SourceFile {
    pub name: PathBuf,
    pub src: Rc<String>,
    pub file_id: FileId,
}

impl SourceFile {
    pub fn new(name: PathBuf, src: String) -> Self {
        let file_id = FileId::from_file_name(&name);

        SourceFile { name, src: Rc::new(src), file_id }
    }
}

#[derive(Debug, Default)]
pub struct FileId(u32);

impl FileId {
    pub fn from_file_name(filename: &Path) -> FileId {
        let mut hasher = DefaultHasher::new();
        filename.hash(&mut hasher);
        FileId(hasher.finish())
    }
}