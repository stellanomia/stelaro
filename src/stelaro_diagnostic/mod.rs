use std::{cell::RefCell, rc::Rc};

use diag::{DiagCtxtHandle, DiagCtxtInner};

pub mod diag;

#[derive(Debug)]
pub struct DiagCtxt {
    inner: RefCell<DiagCtxtInner>,
}

impl DiagCtxt {
    pub fn new(src: Rc<String>) -> DiagCtxt {
        DiagCtxt {
            inner: RefCell::new(
                DiagCtxtInner::new(src)
            ),
        }
    }

    pub fn handle(&self) -> DiagCtxtHandle<'_> {
        DiagCtxtHandle {
            dcx: self,
        }
    }
}
