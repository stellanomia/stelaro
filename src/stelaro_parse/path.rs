use crate::stelaro_ast::{ast::{Path, PathSegment}, token::TokenKind};

use super::{parser::Parser, PResult};


impl<'sess> Parser<'sess> {
    pub fn parse_path(&mut self) -> PResult<Path> {
        let mut segments = vec![];
        let start_span = self.token.span;

        segments.push(
            PathSegment {
                ident: self.parse_ident()?,
                id: self.next_node_id(),
            }
        );

        while self.token.kind == TokenKind::PathSep {
            self.bump();

            segments.push(
                PathSegment {
                    ident: self.parse_ident()?,
                    id: self.next_node_id(),
                }
            );
        }


        Ok(
            Path {
                span: start_span.merge(&self.prev_token.span),
                segments
            }
        )
    }
}
