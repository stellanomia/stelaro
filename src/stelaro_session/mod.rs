pub mod config;
pub mod parse;
pub mod session;

pub use parse::ParseSess;
pub use session::Session;