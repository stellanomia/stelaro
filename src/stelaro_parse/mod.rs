use std::path::Path;

use parser::Parser;

use crate::{stelaro_diagnostic::diag::ErrorEmitted, stelaro_lexer::Lexer, stelaro_session::Session};

pub mod parser;
mod diagnostics;
mod expr;

type PResult<T> = Result<T, ErrorEmitted>;

pub fn new_parser_from_file<'sess>(sess: &'sess Session, path: &Path) -> Result<Parser<'sess>, ErrorEmitted> {
    let file = sess.source_map().load_file(path).unwrap_or_else(|e| {
        sess.dcx().emit_fatal(format!("{e}"));
    });

    let mut lexer = Lexer::new(sess, file.src.as_ref(),);

    match lexer.lex() {
        Ok(ts) => Ok(Parser::new(sess, ts)),
        Err(errs) => Err(errs)?,
    }
}

#[cfg(test)]
mod tests {
    use std::rc::Rc;

    use crate::{stelaro_common::source_map::SourceMap, stelaro_diagnostic::DiagCtxt};

    use super::*;

    #[test]
    fn test_parse_expr() {
        let src = Rc::new(
            "x = (1 + 2) * 3 == 4 and 5 == 6 or 7 != 8 or 9 == 10 and true".to_string()
        );
        let dcx = DiagCtxt::new(Rc::clone(&src));
        let source_map = Rc::new(SourceMap::new());
        let sess = &Session::new(dcx, source_map);
        let mut lexer = Lexer::new(sess, src.as_str());
        let ts = lexer.lex().unwrap();
        let mut parser = Parser::new(sess, ts);
        parser.parse_expr().unwrap();
    }
}