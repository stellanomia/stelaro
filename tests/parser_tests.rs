use std::fs;
use std::path::Path;
use std::rc::Rc;

use stelaro::stelaro_session::Session;
use stelaro::stelaro_parse::{new_parser_from_src, parser::Parser};
use stelaro::stelaro_diagnostic::DiagCtxt;
use stelaro::stelaro_common::source_map::SourceMap;


fn create_test_session(src: Rc<String>) -> Session {
    let dcx = DiagCtxt::new(Rc::clone(&src));
    let source_map = Rc::new(SourceMap::new());
    Session::new(dcx, source_map)
}

fn create_test_parser<'a>(sess: &'a Session, src_str: &'a str) -> Parser<'a> {
    new_parser_from_src(sess, src_str.to_owned()).unwrap()
}


fn run_parser_test(path: &Path) {
    let source_code = fs::read_to_string(path)
        .unwrap_or_else(|e| panic!("テストファイルを読み込むことができませんでした {:?}: {}", path, e));
    let src = Rc::new(source_code);

    let sess = create_test_session(Rc::clone(&src));
    let mut parser = create_test_parser(&sess, &src);

    let parse_result = parser.parse_stelo().unwrap();

    let snapshot_name = path.file_stem() // ファイル名部分を取得 (Option<OsStr>)
        .and_then(|stem| stem.to_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| panic!("ファイル名からスナップショット名を生成できませんでした: {:?}", path));

    insta::assert_debug_snapshot!(snapshot_name, parse_result);
}

#[test]
fn test_parser_inputs() {
    insta::glob!("parser_inputs/*.stelo", |path| {
        run_parser_test(path);
    });
}