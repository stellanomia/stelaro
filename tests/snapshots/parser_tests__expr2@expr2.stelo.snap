---
source: tests/parser_tests.rs
expression: parse_result
input_file: tests/parser_inputs/expr2.stelo
---
Stelo {
    items: [
        Item {
            kind: Fn(
                Function {
                    span: Span {
                        start: 0,
                        end: 64,
                    },
                    sig: FnSig {
                        params: [],
                        ret_ty: Default,
                        span: Span {
                            start: 3,
                            end: 9,
                        },
                    },
                    body: Block {
                        id: NodeId(
                            13,
                        ),
                        stmts: [
                            Stmt {
                                id: NodeId(
                                    12,
                                ),
                                kind: Let(
                                    Local {
                                        id: NodeId(
                                            11,
                                        ),
                                        pat: Pat {
                                            id: NodeId(
                                                1,
                                            ),
                                            kind: Ident(
                                                Ident {
                                                    name: Symbol([ID]),
                                                    span: Span {
                                                        start: 20,
                                                        end: 21,
                                                    },
                                                },
                                            ),
                                            span: Span {
                                                start: 20,
                                                end: 21,
                                            },
                                        },
                                        kind: Init(
                                            Expr {
                                                id: NodeId(
                                                    10,
                                                ),
                                                kind: Block(
                                                    Block {
                                                        id: NodeId(
                                                            9,
                                                        ),
                                                        stmts: [
                                                            Stmt {
                                                                id: NodeId(
                                                                    5,
                                                                ),
                                                                kind: Let(
                                                                    Local {
                                                                        id: NodeId(
                                                                            4,
                                                                        ),
                                                                        pat: Pat {
                                                                            id: NodeId(
                                                                                2,
                                                                            ),
                                                                            kind: Ident(
                                                                                Ident {
                                                                                    name: Symbol([ID]),
                                                                                    span: Span {
                                                                                        start: 38,
                                                                                        end: 39,
                                                                                    },
                                                                                },
                                                                            ),
                                                                            span: Span {
                                                                                start: 38,
                                                                                end: 39,
                                                                            },
                                                                        },
                                                                        kind: Init(
                                                                            Expr {
                                                                                id: NodeId(
                                                                                    3,
                                                                                ),
                                                                                kind: Lit(
                                                                                    Lit {
                                                                                        kind: Integer,
                                                                                        symbol: Symbol([ID]),
                                                                                    },
                                                                                ),
                                                                                span: Span {
                                                                                    start: 42,
                                                                                    end: 44,
                                                                                },
                                                                            },
                                                                        ),
                                                                        ty: None,
                                                                        span: Span {
                                                                            start: 34,
                                                                            end: 45,
                                                                        },
                                                                    },
                                                                ),
                                                                span: Span {
                                                                    start: 34,
                                                                    end: 45,
                                                                },
                                                            },
                                                            Stmt {
                                                                id: NodeId(
                                                                    8,
                                                                ),
                                                                kind: Expr(
                                                                    Expr {
                                                                        id: NodeId(
                                                                            7,
                                                                        ),
                                                                        kind: Path(
                                                                            Path {
                                                                                span: Span {
                                                                                    start: 54,
                                                                                    end: 55,
                                                                                },
                                                                                segments: [
                                                                                    PathSegment {
                                                                                        ident: Ident {
                                                                                            name: Symbol([ID]),
                                                                                            span: Span {
                                                                                                start: 54,
                                                                                                end: 55,
                                                                                            },
                                                                                        },
                                                                                        id: NodeId(
                                                                                            6,
                                                                                        ),
                                                                                    },
                                                                                ],
                                                                            },
                                                                        ),
                                                                        span: Span {
                                                                            start: 54,
                                                                            end: 55,
                                                                        },
                                                                    },
                                                                ),
                                                                span: Span {
                                                                    start: 54,
                                                                    end: 55,
                                                                },
                                                            },
                                                        ],
                                                        span: Span {
                                                            start: 24,
                                                            end: 61,
                                                        },
                                                    },
                                                ),
                                                span: Span {
                                                    start: 24,
                                                    end: 61,
                                                },
                                            },
                                        ),
                                        ty: None,
                                        span: Span {
                                            start: 16,
                                            end: 62,
                                        },
                                    },
                                ),
                                span: Span {
                                    start: 16,
                                    end: 62,
                                },
                            },
                        ],
                        span: Span {
                            start: 10,
                            end: 64,
                        },
                    },
                },
            ),
            id: NodeId(
                14,
            ),
            span: Span {
                start: 0,
                end: 64,
            },
            ident: Ident {
                name: Symbol([ID]),
                span: Span {
                    start: 3,
                    end: 7,
                },
            },
        },
    ],
    span: ModSpan {
        inner_span: Span {
            start: 0,
            end: 64,
        },
    },
    id: NodeId(
        0,
    ),
}
